[NativeAOT/ARM] Assert failure in PutThumb2BlRel24

Crashes when compiling System.Net.Quic.Functional.Tests library tests:

```
[/home/navara/runtime/src/libraries/System.Net.Quic/tests/FunctionalTests/System.Net.Quic.Functional.Tests.csproj::TargetFramework=net9.0-linux]
  Process terminated. Assertion failed.
     at ILCompiler.DependencyAnalysis.Relocation.PutThumb2BlRel24(UInt16* p, Int32 imm24) in /home/navara/runtime/src/coreclr/tools/Common/Compiler/DependencyAnalysis/Relocation.cs:line 177
     at ILCompiler.DependencyAnalysis.Relocation.WriteValue(RelocType relocType, Void* location, Int64 value) in /home/navara/runtime/src/coreclr/tools/Common/Compiler/DependencyAnalysis/Relocation.cs:line 499
     at ILCompiler.ObjectWriter.ObjectWriter.EmitOrResolveRelocation(Int32 sectionIndex, Int64 offset, Span`1 data, RelocType relocType, String symbolName, Int64 addend) in /home/navara/runtime/src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/ObjectWriter/ObjectWriter.cs:line 185
     at ILCompiler.ObjectWriter.ObjectWriter.EmitObject(String objectFilePath, IReadOnlyCollection`1 nodes, IObjectDumper dumper, Logger logger) in /home/navara/runtime/src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/ObjectWriter/ObjectWriter.cs:line 443
     at ILCompiler.ObjectWriter.ObjectWriter.EmitObject(String objectFilePath, IReadOnlyCollection`1 nodes, NodeFactory factory, ObjectWritingOptions options, IObjectDumper dumper, Logger logger) in /home/navara/runtime/src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/ObjectWriter/ObjectWriter.cs:line 533
     at ILCompiler.RyuJitCompilation.CompileInternal(String outputFile, ObjectDumper dumper) in /home/navara/runtime/src/coreclr/tools/aot/ILCompiler.RyuJit/Compiler/RyuJitCompilation.cs:line 111
     at ILCompiler.Compilation.ILCompiler.ICompilation.Compile(String outputFile, ObjectDumper dumper) in /home/navara/runtime/src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/Compilation.cs:line 528
     at ILCompiler.Program.Run() in /home/navara/runtime/src/coreclr/tools/aot/ILCompiler/Program.cs:line 585
     at ILCompiler.ILCompilerRootCommand.<>c__DisplayClass236_0.<.ctor>b__0(ParseResult result) in /home/navara/runtime/src/coreclr/tools/aot/ILCompiler/ILCompilerRootCommand.cs:line 292
     at System.CommandLine.Invocation.InvocationPipeline.Invoke(ParseResult parseResult)
     at System.CommandLine.ParseResult.Invoke()
     at ILCompiler.Program.Main(String[] args) in /home/navara/runtime/src/coreclr/tools/aot/ILCompiler/Program.cs:line 753
  Aborted
/home/navara/runtime/artifacts/bin/coreclr/linux.arm.Debug/build/Microsoft.NETCore.Native.targets(310,5): error MSB3073: The command ""/home/navara/runtime/artifacts/bin/coreclr/linux.arm.Debug/x64/ilc/ilc" @"/home/navara/runtime/artifacts/obj/System.Net.Http.Functional.Tests/Debug/net9.0-linux/native/System.Net.Http.Functional.Tests.ilc.rsp"" exited with code 134. 
```

--- a/src/coreclr/tools/Common/Compiler/DependencyAnalysis/Relocation.cs
+++ b/src/coreclr/tools/Common/Compiler/DependencyAnalysis/Relocation.cs
@@ -163,7 +163,7 @@ namespace ILCompiler.DependencyAnalysis
         //*****************************************************************************
         // Returns whether the offset fits into bl instruction
         //*****************************************************************************
-        private static bool FitsInThumb2BlRel24(int imm24)
+        public static bool FitsInThumb2BlRel24(int imm24)
         {
             return ((imm24 << 7) >> 7) == imm24;
         }

--- a/src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/ObjectWriter/ObjectWriter.cs
+++ b/src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/ObjectWriter/ObjectWriter.cs
@@ -169,8 +169,9 @@ namespace ILCompiler.ObjectWriter
                     // For R_ARM_THM_JUMP24 the thumb bit cannot be encoded, so mask it out.
                     long maskThumbBitOut = relocType is IMAGE_REL_BASED_THUMB_BRANCH24 or IMAGE_REL_BASED_THUMB_MOV32_PCREL ? 1 : 0;
                     long maskThumbBitIn = relocType is IMAGE_REL_BASED_THUMB_MOV32_PCREL ? 1 : 0;
+                    long adjustedAddend = addend;
 
-                    addend -= relocType switch
+                    adjustedAddend -= relocType switch
                     {
                         IMAGE_REL_BASED_REL32 => 4,
                         IMAGE_REL_BASED_THUMB_BRANCH24 => 4,
@@ -178,11 +179,19 @@ namespace ILCompiler.ObjectWriter
                         _ => 0
                     };
 
-                    addend += definedSymbol.Value & ~maskThumbBitOut;
-                    addend += Relocation.ReadValue(relocType, (void*)pData);
-                    addend |= definedSymbol.Value & maskThumbBitIn;
-                    addend -= offset;
-                    Relocation.WriteValue(relocType, (void*)pData, addend);
+                    adjustedAddend += definedSymbol.Value & ~maskThumbBitOut;
+                    adjustedAddend += Relocation.ReadValue(relocType, (void*)pData);
+                    adjustedAddend |= definedSymbol.Value & maskThumbBitIn;
+                    adjustedAddend -= offset;
+
+                    if (relocType is IMAGE_REL_BASED_THUMB_BRANCH24 && (uint)adjustedAddend >= 0xFFFFFF)
+                    {
+                        EmitRelocation(sectionIndex, offset, data, relocType, symbolName, addend);
+                    }
+                    else
+                    {
+                        Relocation.WriteValue(relocType, (void*)pData, adjustedAddend);
+                    }
                 }
             }
             else

--- a/src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/ObjectWriter/ObjectWriter.cs
+++ b/src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/ObjectWriter/ObjectWriter.cs
@@ -184,7 +184,7 @@ namespace ILCompiler.ObjectWriter
                     adjustedAddend |= definedSymbol.Value & maskThumbBitIn;
                     adjustedAddend -= offset;
 
-                    if (relocType is IMAGE_REL_BASED_THUMB_BRANCH24 && (uint)adjustedAddend >= 0xFFFFFF)
+                    if (relocType is IMAGE_REL_BASED_THUMB_BRANCH24 && !Relocation.FitsInThumb2BlRel24((int)adjustedAddend))
                     {
                         EmitRelocation(sectionIndex, offset, data, relocType, symbolName, addend);
                     }

